src/core_plugins/metrics/public/components/vis_editor.js

```diff
@@ -60,27 +69,67 @@ class VisEditor extends Component {
     }
   }
 
-  render() {
-    const handleChange = (part) => {
-      const nextModel = { ...this.state.model, ...part };
-
-      this.props.vis.params = nextModel;
-      if (this.state.autoApply) {
-        this.props.vis.updateState();
-      }
+  isIndexPatternChanged = () => {
+    if (this.state.model.index_pattern !== this.props.vis.params.index_pattern) {
+      return true;
+    }
+    const visAnnotations = this.props.vis.params.annotations || [];
+    const stateAnnotations = this.state.model.annotations || [];
+    const currentAnnotations = new Set(visAnnotations.map(annotation => annotation.index_pattern));
+    const prevAnnotations = new Set(stateAnnotations.map(annotation => annotation.index_pattern));
+    const diff = [...currentAnnotations].filter(a => !prevAnnotations.has(a));
+    return diff.length !== 0;
+  }
 
-      this.setState({ model: nextModel, dirty: !this.state.autoApply });
-    };
+  async fetchIndexPatternFields(vis) {
+    if (vis.params.index_pattern === '') {
+      // set the default index pattern if none is defined.
+      const savedObjectsClient = chrome.getSavedObjectsClient();
+      const indexPattern = await savedObjectsClient.get('index-pattern', this.getConfig('defaultIndex'));
+      const defaultIndexPattern = indexPattern.attributes.title;
+      vis.params.index_pattern = defaultIndexPattern;
+    }
+    const indexPatterns = extractIndexPatterns(vis);
+    const fields = await fetchFields(indexPatterns);
+    return fields;
+  }
 
-    const handleAutoApplyToggle = (part) => {
-      this.setState({ autoApply: part.target.checked });
-    };
+  checkIndexPatternChanges = async (forceFetch) => {
+    if (forceFetch || this.isIndexPatternChanged()) {
+      const fields = await this.fetchIndexPatternFields(this.props.vis);
+      const visFields = {
+        ...this.state.visFields,
+        ...fields,
+      };
+      this.setState({
+        visFields,
+      });
+    }
+  }
 
-    const handleCommit = () => {
+  handleChange = async (part) => {
```

Comment:
nit: can we rename `part` to something that makes it easier to read ?

Reply:
sure, I will rename this, was just moved from the previous code